### PR TITLE
Add blog post “sustainable scala”

### PR DESCRIPTION
**Pull request by julienrf**
May I ask a review? :pray: 

Screenshot 2021-12-13 at 16-27-32 Sustainable Scala

### CODE REVIEW

File: blog/_posts/2021-12-14-sustainable-scala.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+|---|---|---|
+| C | 1.00 | 0.00 |
+| Java | 2.04 | 1.45 |
+| **Scala** | **3.71** | **3.67** |
+| **Scala (idiomatic)** | **6.99** | **4.87** |
+| JavaScript | 7.63 | 3.62 |
+| Python | 89.33 | 113.79 |
+
+A similar table was shown by Rui Pereira _et. al_ [^4]. While we did not get 
+exactly the same numbers, the order of magnitudes remain the same.
```
Comment (SethTisue) [inline blog/_posts/2021-12-14-sustainable-scala.md:228]: s/order of magnitudes/orders of magnitude/

File: blog/_posts/2021-12-14-sustainable-scala.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+instance, the energy consumption of the `k-nucleotide` benchmark was between 
+2 to 13 times higher than the C implementation. The differences between both 
+versions are mainly the usage of immutable Scala collections and `for` loops in 
+the idiomatic version, as opposed to `Array`s and `while` loops in the 
+optimized version (this observation is consistent with the results of 
+another study ran by Rui Pereira _et. al._ [^5]).
+
+We believe that the fact that Scala embraces several programming paradigms 
+is a strength. It makes it easy to write high-level code that reads well, 
+and it also makes it easy to write low-level that performs well.
```
Comment (SethTisue) [inline blog/_posts/2021-12-14-sustainable-scala.md:245]: s/low-level that/low-level code that/

**Review by SethTisue — APPROVED** [review]
This was really interesting reading.

One thing I'm wondering is:

I assume that in general, the longer a program takes to run, the more energy it's consuming. You write:

> They compared the execution time, energy consumption, and memory consumption [...]

Did they ever find that execution time and energy consumption _didn't_ simply vary in tandem? If so, then: in what way, under what circumstances?

Anyway, this is just something I was personally curious about. If you think other readers might be curious about it as well, you might consider answering it in the post itself.

File: blog/_posts/2021-12-14-sustainable-scala.md (hunk 1 above)
Comment (SethTisue) [reply blog/_posts/2021-12-14-sustainable-scala.md:228]: (in the spirit of https://www.theonion.com/william-safire-orders-two-whoppers-junior-1819565735)

**Comment by julienrf** [comment]
> Did they ever find that execution time and energy consumption didn't simply vary in tandem? If so, then: in what way, under what circumstances?

The energy consumption is proportional to the execution time: Energy (J) = Time (s) × Power (W). However, it seems (based on the measurements) that the power is not constant, so the assumption “the faster, the greener” may be too simplistic, although it is often true.

File: blog/_posts/2021-12-14-sustainable-scala.md (hunk 1 above)
Comment (julienrf) [reply blog/_posts/2021-12-14-sustainable-scala.md]: Thank you! Fixed in 74171c26

File: blog/_posts/2021-12-14-sustainable-scala.md (hunk 2 above)
Comment (julienrf) [reply blog/_posts/2021-12-14-sustainable-scala.md]: Thank you! Fixed in 74171c26

**Comment by joroKr21** [comment]
> I assume that in general, the longer a program takes to run, the more energy it's consuming. You write:

I think in benchmarks this might be true, but not necessarily in real world code. Think about async vs blocking database access. The async one might actually take a bit longer, but it consumes less resources (not blocking a thread) and intuitively it should also spend less energy per transaction.

File: blog/_posts/2021-12-14-sustainable-scala.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+A similar table was shown by Rui Pereira _et. al_ [^4]. While we did not get 
+exactly the same numbers, the orders of magnitude remain the same.
+
+We now have an answer to our initial question, “where does Scala 
+stand in the picture?” According to these benchmarks, Scala is well positioned 
+within high-level programming languages.
+
+Also, we see that in Scala, two implementations of the same benchmark can 
+easily show different performance, depending on your code style. For
```
Comment (vincenzobaz) [inline blog/_posts/2021-12-14-sustainable-scala.md:235]:
```suggestion
easily show different performances, depending on your code style. For 
```

**Review by vincenzobaz — APPROVED** [review]
Great article!

**Comment by SethTisue** [comment]
I put this up on Reddit at https://www.reddit.com/r/scala/comments/rgj6i1/sustainable_scala_blog_post_by_julien_richardfoy/

@julienrf I assume scala_lang will tweet it out...?

**Pull request merged**